README

Aufs -- Another Unionfs
Junjiro Okajima

# $Id: README,v 1.46 2007/04/30 05:48:24 sfjro Exp $


0. Introduction
----------------------------------------
In the early days, aufs was entirely re-designed and re-implemented
Unionfs Version 1.x series. After many original ideas, approaches,
improvements and implementations, it becomes totally different from
Unionfs while keeping the basic features.
Unionfs is being developed by Professor Erez Zadok at Stony Brook
University and his team.
If you don't know Unionfs, I recommend you to try and know it before
using aufs. Some terminology in aufs follows Unionfs's.

Bug reports (including my broken English), suggestions, comments
and donations are always welcome. Your bug report may help other users,
including future users. Especially the bug which doesn't follow unix/linux
filesystem's semantics is important.


1. Features
----------------------------------------
- unite several directories into a single virtual filesystem. The member
  directory is called as a branch.
- you can specify the permission flags to the branch, which are 'readonly',
  'readwrite' and 'whiteout-able.'
- by upper writable branch, internal copyup and whiteout, files/dirs on
  readonly branch are modifiable logically.
- dynamic branch manipulation, add, del.
- etc... see Unionfs in detail.

Also there are many enhancements in aufs, such like,,,
- safer and faster
- keep inode number by external inode number table
- keep the timestamps of file/dir in internal copyup operation
- seekable directory, supporting NFS readdir.
- support mmap(2) including /proc/PID/exe symlink, without page-copy
- whiteout is hardlinked in order to reduce the consumption of inodes
  on branch
- do not copyup, nor create a whiteout when it is unnecessary
- revert a single systemcall when an error occurs in aufs
- remount interface instead of ioctl
- maintain /etc/mtab by an external shell script, /sbin/mount.aufs.
- loopback mounted filesystem as a branch
- kernel thread for removing the dir who has a plenty of whiteouts
- support copyup sparse file (a file which has a 'hole' in it)
- default permission flags for branches
- selectable permission flags for ro branch, whether whiteout can
  exist or not
- export via NFS.
- <sysfs>/fs/aufs.
- and more... see aufs manual in detail

Aufs is in still development stage, especially,,,
- pseudo hardlink (hardlink over branches)
- allow a direct access manually to a file on branch, e.g. bypassing aufs.
  including NFS or remote filesystem branch.

(current work)
- refine xino and revalidate
- lockdep
- a delegation of the internal branch access to support task I/O
  accounting, which also supports Linux Security Modules (LSM) mainly
  for Suse AppArmor.
- nested mount, i.e. aufs as readonly no-whiteout branch of another aufs.

(just an idea)
- sample for kvm, xen, vmware and openvz? If a directory can be
  sharable between several Guest-OSes, to share readonly aufs branches
  will be an advantage.
- remount option copy/move between two branches. (unnecessary?)
- O_DIRECT (unnecessary?)
- light version, without branch manipulation. (unnecessary?)
- SMP, because I don't have such machine. But several users reported
  aufs is working fine on SMP machines.
- 64bit environment, because I don't have such machine.
- copyup in userspace
- xattr, acl
- disk quota (unnecessary?)
- and documentation


2. Download
----------------------------------------
CVS tree is in aufs project of SourceForge.
Here is simple instructions to get aufs source files. It is recommended to
refer to the document about CVS on SourceForge.
	$ mkdir aufs.wcvs
	$ cd aufs.wcvs
	$ cvs -d:pserver:anonymous@aufs.cvs.sourceforge.net:/cvsroot/aufs login
	(CVS password is empty)
	$ cvs -z3 -d:pserver:anonymous@aufs.cvs.sourceforge.net:/cvsroot/aufs co aufs

In order to update files after the first checkout.
	$ cd aufs.wcvs/aufs
	$ cvs update

In order to see what the difference between two versions (two dates) is,
	$ cd aufs.wcvs/aufs
	$ cvs diff -D20061212 -D20061219

I always try putting the stable version in CVS, so you can try CVS
instead of SourceForge File Release. And every changes are summarized
and reported to aufs-users at lists.sourceforge.net ML. It is recommended to
join this ML.


3. Configuration and Compilation
----------------------------------------
Aufs is being developed and tested on linux-2.6.16 and later.
You need the correct environment to build a kernel module. Generally
it is an environment which your kernel was built.
If you are unsure that you got the correct environment, then I
recommend you to built your kernel by yourself. If it succeeds, you've
got the correct environment.

Currently aufs configuration is written in ./Kconfig.in. You can build the
real ./fs/aufs/Kconfig file by 'make.' And follow the instructions which will
be produced. The built ./fs/aufs/Kconfig depends upon the kernel version.
	$ make -f local.mk kconfig
The local.mk searches your kernel build path by
	KDIR = /lib/modules/$(shell uname -r)/build
If you are cross-compiling the aufs module, try
	$ make KDIR=/your/kernel/build/path -f local.mk kconfig
If you compile aufs statically, the built Kconfig will help you.

Also you can use ./local.mk to compile aufs as a module by simply,
	$ make -f local.mk
or
	$ make KDIR=/your/kernel/build/path -f local.mk

The default configuration is written in ./local.mk too, they will work
for you in most cases. You can edit this file and change the
configuration for your aufs module.
If you don't want to change the ./local.mk, then create a new file
./priv_def.mk and write the definitions your aufs configuration. The
./local.mk includes ./priv_def.mk it it exists.

Currently, CONFIG_DEBUG_PROVE_LOCKING (in linux kernel) is not
supported since MAX_LOCKDEP_SUBCLASSES is too low for a stackable
filesystem. I am trying reducing the depth of lock subclasses.

When you test the performance, it is recommended to disable
CONFIG_AUFS_DEBUG. It is enabled by default for the feature test.


4. Usage
----------------------------------------
After 'make',
	$ man -l ./aufs.5
	# install -m 500 -p mount.aufs umount.aufs auplink aulchown /sbin (recommended)
	# insmod ./aufs.ko
	$ mkdir /tmp/rw /tmp/aufs
	# mount -t aufs -o dirs=/tmp/rw:${HOME}=ro none /tmp/aufs

Here is another example.

	# mount -t aufs -o br:/tmp/rw:${HOME}=ro none /tmp/aufs
	or
	# mount -t aufs -o br:/tmp/rw none /tmp/aufs
	# mount -o remount,append:${HOME}=ro /tmp/aufs

If you disable CONFIG_AUFS_COMPAT in your configuration, you can remove the
default branch permission '=ro.'

	# mount -t aufs -o br:/tmp/rw:${HOME} none /tmp/aufs

Then, you can see whole tree of your home dir through /tmp/aufs. If
you modify a file under /tmp/aufs, the one on your home directory is
not affected, but the same named file will be newly created under
/tmp/rw. And all of your modification to the file will be applied to
the one under /tmp/rw. This is called Copy on Write (COW) method.
Aufs mount options are described in the generated ./aufs.5.

Additionally, there are some sample usages of aufs which are a
diskless system with network booting, and LiveCD over NFS.
See ./sample/diskless in detail.


5. Contact
----------------------------------------
When you have any problems or strange behaviour in aufs, please let me
know with,,,
- /proc/mounts (instead of the output of mount(8))
- /sys/fs/aufs/* (if you have them)
- linux kernel version
- AUFS_VERSION (defined in include/linux/aufs_type.h)
- configuration (define/undefine CONFIG_AUFS_xxx, or plain local.mk is
  used or not)
- phenomenon
- actual operation, reproducible one is better
- mailto: aufs-users at lists.sourceforge.net

Usually, I don't watch the Public Areas(Bugs, Support Requests, Patches,
Feature Requests) on SourceForge. Please join and write to aufs-users ML.


6.
----------------------------------------
If you are an experienced user, no explanation is needed. Aufs is
just a linux filesystem module. take a glance at ./local.mk,
aufs.5, and Unionfs.


Enjoy!

# Local variables: ;
# mode: text;
# End: ;


History

History of aufs
Junjiro Okajima

# $Id: History,v 1.18 2007/04/30 06:01:09 sfjro Exp $

Note:
- the described date are all in JST.
- CVS handles UTC as its timezone.
- all these descriptions should be archived in the ML on sourceforge,
  since I always post there and let users know. But sometimes archiving
  failed. this is the main reason to begin writing this file.

----------------------------------------------------------------------

20070430
- bugfix: re-use inode->i_generation to support the enexpectedly
  survived inodes after remount. test lower inodes in
  hidden_d_revalidate(). simplified postproce() in
  udba=inotify. introduce an internal generation for aufs_iinfo.
  reported by Tomas M.
  it is not finished yet and a hardlink on NFS branch will not work.
- bugfix: at deleting a branch, revalidate the survived dentries after
  shrink dcache.
- bugfix: support the unhashed dentries, but ignore the unreferenced
  dentries.
- bugfix: ii_read/write_unlock in an error path, mainly for NFS
  branch.
- bugfix: revalidating the opened files after remount, introduce a new
  function au_hfput().
- bugfix: maintain the reference of aufs module in case of the no-wait
  task.
- introduce <sysfs>/fs/aufs/<mnt_id>/xino, and sysaufs.h. move some
  declarations from module.h to sysaufs.h. rename del/add_sbilist() to
  sysaufs_del/add(). refine sysaufs.c simply.
- CONFIG_AUFS_EXPORT in statically linked aufs depends upon
  CONFIG_EXPORTFS=y, reported by Sandino Flores Moreno.
- call dget_parent() instead of accessing denrty->d_parent directly.
- convert some inline functions for debugging to macros, in order to
  print the prefeable line number in a file.
- rename nfsd_lockdep_off/on() to au_...().
- supports IN_UNMOUNT, if it happens there must must be another bug.
- ignore whiteouted names in revalidating child files.
- rename init_export_op() to au_init_export_op().

20070423
- bugfix: file revalidation after rename(tgt), reported by Just Marc.
- strict inode lock in fsync() before 2.6.17.
- restore rdacahe option, which was dropped by mistake.
- print the depth of generic workqueue in sysfs/aufs/stat.
- testing initramfs.
- still testing udba=inotify.
- simplified "uid/gid/perm" message.
- reduce stack space.
- make aufs_cond available even if CONFIG_LKTR is disabled.
- revise au_fin_finfo() and si lock.
- simplified cache function definitions.
- merge both of wait/nowait works into a single enqueue function.
- rename wkq_wait() to au_wkq_wait().
- rename some name hash functions, and two new functions nash_new()
  and nhash_del().
	+ init_nhash()	--> nhash_init()
	+ move_nhash()	--> nhash_move()
	+ free_nhash()	--> nhash_fin()
- move au_is_kthread() to wkq.h, and rename to is_aufsd().
- rename au_init_wkq() to au_wkq_init().
- rename au_fin_wkq() to au_wkq_fin().
- rename au_inotify_exit() to au_inotify_fin().
- rename dbg_dlgt_exit() to dbg_dlgt_fin().

20070416
- bugfix: introduce dcsub.[ch] to handle dentry cache.
- bugfix: use readlink(1) to get cwd, in auplink, mount.aufs and
  unionctl, reported by Torsten Luettgert. Actually, to use readlink
  was suggested by Tomas M last year.
- support initramfs.
- verbose warning message for 'different uid/gid/perm.'
- still testing udba=inotify.
- support GFP flags in au_kzrealloc().
- simplified lock/unlock functions definition by a new macro
  SimpleRWSem...

20070409
- introduce new branch attribute "nolwh."
- refine parsing options at mount/remount time, reported by Sandino
  Flores Moreno.
	+ remove whiteout-base and whplink-dir in nopllink and br_mod(),
	  but br_del().
	+ increment sbinfo generation in a new function
	  aufs_remount_fs().
	+ support multiple xino/noxino options in another way.
	+ force "ro" when unknown branch permission is specified.
	+ at mount-time, interpret options in this order.
		* simple flags, except xino/noxino, udba=inotify and dlgt.
		* branches.
		* xino/noxino.
		* udba=inotify.
		* dlgt.
	+ at remount-time,
		* interpret options in the given order left to right,
		  except dlgt which is disabled in interpreting.
		* create whiteout-base and whplink-dir if neccessary.
		* dlgt if neccessary.
	+ new functions, clean_wh(), clean_plink(), test_linkable(),
	  plink_dir().
- support 64bit environtment, reported by Vasily Tarasov.
- introduce a sample for kprobe.
- bugfix: do not traverse the sub-mounts under aufs at refreshing
  cached dirs after remount.
- bugfix: simultaneous read in 'dlgt' mode, reported and fixed by
  Ulrich Holeschak.
- bugfix: LINUX_VERSION_CODE for linux-2.6.16.
- still testing inotify.
- support lockdep in nfsd, still testing.
- minor optimization for dput().
- new test functions br_writable(), br_whable() and br_linkable_wh(),
  replacing au_is_whable().
- do write_lock() in reinit_br_wh().
- refine updating /sys/fs/aufs/stat.
- wkq_nowait() calls schedule_work().

20070402
- bugfix: forgot dput in a error case.
- introduce wkq_nowait().
- refine udba=inotify, still testing.
- extract some functions from sbinfo.c, create a new file plink.c.

20070328
- bugfix: forgot include one header in misc.h

20070327
- bugfix: increment sbinfo generation in cases of br_mod() for sensing
  whiteout, xino/noxino and udba=inotify, too. introducing
  au_is_whable() and sbr_is_whable(). call au_sigen_inc() from
  do_opts(), reported by Jorgen P. Tjerno.
- support sysfs and NFS-exporting, introducing sysaufs.c,
  CONFIG_AUFS_SYSAUFS, export.c, CONFIG_AUFS_EXPORT and a new module
  parameter 'brs.'
- remove AUFS_DEBUG_RWSEM.
- workaround an application bug when 'dlgt' is specified.
  generally, read(2) or write(2) may return the value shorter
  than requested. But many applications don't support it,
  for example bash.
- support file gid when 'dlgt' is set.
- add a note about SMP.
- add a note about lockdep, still testing.
- add a note about performance test.
- convert branch permission bits into an integer.
- add an branch mode alias 'rowh' which is equivalent to 'ro+wh.'

fs/aufs/aufs.h
/*
 * Copyright (C) 2005, 2006, 2007 Junjiro Okajima
 *
 * This program, aufs is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

/* $Id: aufs.h,v 1.22 2007/04/30 05:43:57 sfjro Exp $ */

#ifndef __AUFS_H__
#define __AUFS_H__

#include <linux/version.h>

/* ---------------------------------------------------------------------- */

// limited support before 2.6.16, curretly 2.6.15 only.
#if LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,15)
#define atomic_long_t		atomic_t
#define atomic_long_set		atomic_set
#define timespec_to_ns(ts)	({(long long)(ts)->tv_sec;})
#define D_CHILD			d_child
#else
#define D_CHILD			d_u.d_child
#endif

/* ---------------------------------------------------------------------- */

#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,18)
#define I_MUTEX_QUOTA			0
#define lockdep_off()			/* */
#define lockdep_on()			/* */
#define mutex_lock_nested(mtx, lsc)	mutex_lock(mtx)
#define down_write_nested(rw, lsc)	down_write(rw)
#define down_read_nested(rw, lsc)	down_read(rw)

#ifdef CONFIG_LOCKDEP
# define DECLARE_COMPLETION_ONSTACK(work) \
	struct completion work = COMPLETION_INITIALIZER_ONSTACK(work)
#else
# define DECLARE_COMPLETION_ONSTACK(work) DECLARE_COMPLETION(work)
#endif
#endif

/* ---------------------------------------------------------------------- */

#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,19)
#define filldir_ino_t	u64
#define filldir_ino_fmt	"%Lu"
#else
#define filldir_ino_t	ino_t
#define filldir_ino_fmt	"%lu"
#endif

/* ---------------------------------------------------------------------- */

#include "debug.h"

#include "branch.h"
#include "cpup.h"
#include "dcsub.h"
#include "dentry.h"
#include "dir.h"
#include "file.h"
#include "inode.h"
//#include "kobj.h"
#include "misc.h"
#include "module.h"
#include "opts.h"
#include "super.h"
#include "sysaufs.h"
#include "vfsub.h"
#include "whout.h"
#include "wkq.h"
//#include "xattr.h"

#if defined(CONFIG_AUFS_MODULE) && !defined(CONFIG_AUFS_KSIZE_PATCH)
#define ksize(p)	(-1)
#endif

#endif /* __AUFS_H__ */
